### PR TITLE
Improved Stock Reduction Events

### LINKED ISSUE #5117: Manually Adding Completed Order Doesn't Reduce Stock
Had this through a ticket.

Initial stock: http://cld.wthms.co/N4Dy

Order manually added and completed: http://cld.wthms.co/bcAf

*EDIT*

Ahh, the stock doesn't change unless you select the product and choose reduce line stock. Is this the smartest way to handle it for now? Seems that a completed order, even if added manually, should reduce the stock by the qty set when adding the product to the order.

### LINKED ISSUE #13349: Double stock removal on edit order with payment method change?
Hi,

So I am having a issue where the same order gets stock removed 2 or more times.

Scenario example:

1. Client checkout out simple product and chooses Bank Transfer and finnish.
2. Order goes "on-hold" and stock is removed
3. Client requests new payment method or If by chance I proceed again with bank transfer 
4. I go to edit the order and have the  option "go to client payment page" above the "order status" select
5. I go to "client payment page" select new method or same and proceed.
6. New payment details are created, and again **stock gets removed for the same order**...

Somehting wrong here, aint there a condition in "reduce_order_stock()" that when called before reducing the stock checks if:

1.  Stock for item has already been removed for that order being edited
2. if stock same has before edit than dont do nothing
3. if stock diferent then **add back previous removed stock** and then **remove stock for with the new value**

I didnt find the condition in the programing for this... maybe i missed.

Thanks for the feedback.

image

`
### WordPress Environment ###

WC Version: 2.6.14
Log Directory Writable: ✔
WP Version: 4.7.2
WP Multisite: –
WP Memory Limit: 512 MB
WP Debug Mode: –
WP Cron: –
Language: pt_PT

### Server Environment ###

Server Info: Apache
PHP Version: 5.6.30
PHP Post Max Size: 256 MB
PHP Time Limit: 120
PHP Max Input Vars: 1000
cURL Version: 7.36.0
OpenSSL/1.0.1e

SUHOSIN Installed: –
MySQL Version: 5.6.35
Max Upload Size: 256 MB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ✔
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Post Remoto: ✔
Get Remoto: ✔

### Database ###

WC Database Version: 2.6.14
: 
woocommerce_sessions: ✔
woocommerce_api_keys: ✔
woocommerce_attribute_taxonomies: ✔
woocommerce_downloadable_product_permissions: ✔
woocommerce_order_items: ✔
woocommerce_order_itemmeta: ✔
woocommerce_tax_rates: ✔
woocommerce_tax_rate_locations: ✔
woocommerce_shipping_zones: ✔
woocommerce_shipping_zone_locations: ✔
woocommerce_shipping_zone_methods: ✔
woocommerce_payment_tokens: ✔
woocommerce_payment_tokenmeta: ✔
MaxMind GeoIP Database: ✔

### Active Plugins (34) ###

iThemes Security: por iThemes – 6.1.1
Breadcrumb NavXT: por John Havlik – 5.6.0
Contact Form 7: por Takayuki Miyoshi – 4.6.1
createIT Waitlist Plugin: por createIT – 1.7
WPBakery Visual Composer: por Michael M - WPBakery.com – 5.0.1
MagniFinance Invoice System: por WebDS – 1.0
MEO Wallet Gateway de pagamento: por WebDS – 2.0.1
Products Purchase Price for Woocommerce: por Softedge – 1.0.4
Slider Revolution: por ThemePunch – 5.2.6
WooCommerce Advanced Shipping: por Jeroen Sormani – 1.0.12
WooCommerce - Country Based Payments: por Ivan Paulin – 1.1.6.1
WooCommerce - Store Exporter: por Visser Labs – 1.8.8
WooCommerce POS: por kilbot – 0.4.6
WooCommerce: por WooThemes – 2.6.14
Importador do WordPress: por wordpressdotorg – 0.6.3
Yoast SEO: por Team Yoast – 4.3
WP-Mail-SMTP: por Callum Macdonald – 0.10.1
MailPoet Newsletters: por MailPoet – 2.7.7
YITH WooCommerce Ajax Search: por YITHEMES – 1.4.1
YITH WooCommerce Wishlist: por YITHEMES – 2.0.16
Zendesk Chat Widget: por Zopim – 1.4.6

### Settings ###

Force SSL: –
Currency: EUR (€)
Currency Position: right
Thousand Separator: ,
Decimal Separator: .
Number of Decimals: 2

### API ###

API Enabled: ✔

### WC Pages ###

Loja base: #1818 - /shop/
Carrinho: #2070 - /cart/
Finalizar compras: #2828 - /checkout/
Minha conta: #2829 - /my-account/

### Taxonomies ###

Product Types: external (external)
grouped (grouped)
simple (simple)
variable (variable)

### Theme ###

Name: Oxy Child
Version: 1.0
Author URL: http://www.smartdatasoft.com
Child Theme: ✔
Parent Theme Name: OXY
Parent Theme Version: 2.9
Parent Theme Author URL: http://www.smartdatasoft.com/
WooCommerce Support: ✔

### LINKED ISSUE #13359: Stock Reduction Events
This issue should take care of:

- [x] Manually adding orders does not reduce stock https://github.com/woocommerce/woocommerce/issues/5117
- [x] Order stock may be reduced twice if the order is paid for again using different gateways https://github.com/woocommerce/woocommerce/issues/13349
- [x] If stock reduction is complete (or done manually) we can disable stock level checks on payment. https://github.com/woocommerce/woocommerce/issues/18399

We need to have more robust stock reduction/increase events and ensure it's only done once.

### LINKED ISSUE #18399: Customer cannot pay for seller-prepared orders when there is no stock
The situation: I have some spare parts that I want to sell to a specific customer through the Woocommerce shop (WC 3.2.6, WP 4.9.1).

I do not want to add stock to the shop items because I do not want to risk others buying these items. Hence, I manually create an order for this customer and add the respective items.

The customer receives a link to his order, and clicks "Pay" and wants to pay (installed gateways: Paypal express and Stripe).

Instead of being able to check out, he receives an error message instead, saying that the items are out of stock. This is not the intended behavior in this case.

I went through the effort of increasing the stock for each item, and asking the customer to check out quickly. This worked out well, so there don't seem to be other problems.

### PR DESCRIPTION
**Pull request by mikejolley**
This is in progress. Attempt at resolving #13359 (had this local for a while, finally got around to finishing it).

Basic gist:

- We need to listen to certain events to increase or decrease stock levels for orders
- Stock may be held up by pending orders
- Manual orders can be added by admin users. The stock should be held until payment like regular orders.
- Paying for orders should not error due to stock, if the admin created them.

Changes:

1. I've removed the increase/decrease stock bulk actions in the order - this is so our automation events can take control of this. If it's a pending order, the stock is held, so manual stock reduction should not be necessary.
2. I've added tracking at line item level to know if stock was reduced or not.
3. For manual orders, I've improved the modal so you can set quantities, and included stock levels in the search results for reference:
slack - a8c 2018-06-08 16-59-04
4. In the cart, rather than only 'holding' stock only if the setting is used, all pending orders stock are held. The timeout setting will cancel these orders and free up this stock again.
5. For the page page, stock is only re-checked if the stock was not held (failed status)
6. Refactored some stock functions to make them reusable.
7. Decrease events include payment, order status on-hold/pending/completed
8. Increase events include order cancelled/pending.

The automated stock reduction events resolve #5117
The check for pending orders (to avoid stock checks due to stock already being held) resolves #18399
The additional checks before reducing stock resolve #13349
This PR closes #13359

Looking for some feedback before polishing or fixing any last issues. cc @claudiulodro

### CODE REVIEW

File: includes/wc-stock-functions.php [hunk 1]
```diff
@@ -98,48 +144,175 @@ function wc_reduce_stock_levels( $order_id ) {
...
+		return;
+	}
+
+	// Loop over all items.
+	foreach ( $order->get_items() as $item ) {
+		if ( ! $item->is_type( 'line_item' ) ) {
+			continue;
+		}
+
+		// Only reduce stock once for each item.
```
Comment (claudiulodro) [inline includes/wc-stock-functions.php]: Comment should say 'increase stock once . . .'

**Comment by claudiulodro** [comment]
- Code looks good.
- Manual order stock reduction works well. Order notes are good and help clarify what's going on.
- New "Add Product" modal for orders looks good. The quantity input on it should be a number input instead of a text input. Currently you get a line item with quantity 0 created on non-numeric inputs for that field.
- I noticed some sort of bug when deleting line items. Here is how to reproduce:
	1. Try and delete a line item using the 'x'.
	2. Observe the line item is still rendered but with no picture, cost, or quantity.
	3. Delete the line item again using the 'x'.
	4. Observe the "ghost" line item properly deletes on the 2nd try.
- The warning alert message that appears when you click the 'x' to delete a line item talks about manually adding back reduced stock. Does the alert message need updating?

Overall, this looks great and is very intuitive in the way it works (stock goes up or down when I would naturally expect it to go up or down). Good work. 👍

File: includes/class-wc-ajax.php [hunk 2]
```diff
@@ -1411,7 +1329,14 @@ public static function json_search_products( $term = '', $include_variations = f
 		$products        = array();
 
 		foreach ( $product_objects as $product_object ) {
-			$products[ $product_object->get_id() ] = rawurldecode( $product_object->get_formatted_name() );
+			$formatted_name = $product_object->get_formatted_name();
+			$managing_stock = $product_object->managing_stock();
+
+			if ( $managing_stock ) {
+				$formatted_name .= ' &ndash; ' . wc_format_stock_for_display( $product_object );
```
Comment (franticpsyx) [inline includes/class-wc-ajax.php]: @mikejolley this is very useful when adding items to orders manually, so 👍 for that! 

From what I can tell, the added stock details will be displayed when searching for products in other places, e.g. in Up-Sells / Cross-Sells fields, or in custom search fields using `json_search_products`. Not sure if stock data is relevant in all cases? Might be worth finding a way to control this, either by adding a parameter in the function, or by adding a filter here.

Comment (mikejolley) [reply includes/class-wc-ajax.php]: Yeah I agree. Not sure how yet.

**Review by claudiulodro — APPROVED** [review]
Looks good to me. 👍

**Pull request merged**